pythonRelaxBuildDepsHook: init

nicoonoclaste opened the pull request:
- [x] `pythonRelaxBuildDepsHook`: init
  new setup hooks which rewrites `pyproject.toml`'s `build-system.requires`
  - [x] implementation comments
  - [x] documentation
  - [ ] tests
- [x] `mkPythonDerivation`: add `pythonRelaxBuildDepsHook` when its attributes are set
  analoguous to the existing `pythonRelaxDepsHook`
- [x] `ansible`: cleanup using the new hook

## Rationale

This was first devised in #381523, to replace regex-based `pyproject.toml` rewrites:
- using actual parsers for TOML and PEP 508 dependencies specifications, is a lot more robust ;
- a quick search through nixpkgs finds a huge number of derivations which could be cleaned up this way ;
  there are also many dependencies performing other types of rewrites in `pyproject.toml`,
  but this may best be left to another tool (or extending this hook in a future PR)

Regarding implementation choices:
- `tomlkit` was chosen as it preserves formatting and comments, making diff-based inspection simpler ;
- comments on substituted `build-system.requires` entries aren't preserved, to keep the implementation simpler.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of `pythonRelaxBuildDepsHook`
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nicoonoclaste committed:
pythonPackages.pythonRelaxBuildDepsHook: init

nicoonoclaste committed:
mkPythonDerivation: add `pythonRelaxBuildDepsHook` if the relevant attributes are set

nicoonoclaste committed:
ansible: tidy-up using `pythonRelaxBuildDepsHook`

nbraud commented on pkgs/development/interpreters/python/hooks/default.nix:
```diff
@@ -230,6 +230,29 @@ in
+      {
+        name = "python-relax-build-deps-hook";
+      }
+      (writeTextFile {
+        name = "python-relax-build-deps-hook.sh";
+        text = ''
+          _pythonRelaxBuildDepsHook() {
+            PYTHONPATH="${packaging}/${pythonSitePackages}:${tomlkit}/${pythonSitePackages}:$PYTHONPATH" \
+              ${pythonInterpreter} ${./python-relax-build-deps-hook.py}
+          }
+
+          preConfigureHooks+=(_pythonRelaxBuildDepsHook)
+        '';
+      })
```
This could be plonked in a text file (using `makeHook`'s substitution mechanism) but it seems worse:
anyone reading the code would have to first open `python-relax-build-deps-hook.sh`, only to find out it's a trivial stub and follow to the actual implementation instead.

nbraud commented on pkgs/development/interpreters/python/hooks/python-relax-build-deps-hook.py:
```diff
@@ -0,0 +1,42 @@
+relaxDeps = nixAttrs.get("pythonRelaxBuildDeps", {})
+if isinstance(relaxDeps, list):
+    relaxDeps = { req.name: req for req in map(Requirement, relaxDeps) }
+else:
+    def requirementWithSpec(name, spec):
+        """Construct a Requirement with a given name and version specifier"""
+        r = Requirement(name)
+        r.specifier = SpecifierSet(spec)
+        return r
+
+    relaxDeps = {
+        name: requirementWithSpec(name, spec)
+        for name, spec in relexDeps.items()
+    }
```
While I like the idea of going `pythonRelaxBuildDeps.foo = ">1, <2";` or such, especially if version constraints need to be (in part) programatically generated

... I'm not sure the usecase is there at the moment, and it comes with non-negligible complexity (~20% of the code, and among the less nice-to-read parts)

nbraud replied:
I take that back: the list form permits writing something like `[ "foo =1" "foo =2" ]`, which would be messier to even just find and error-out on.

While the attrset form diverges from `pythonRelaxDeps`' (current) API, it entirely obviate this issue, so I believe this should be the only accepted form instead.

FWIW, it looks like `pythonRelaxDepsHook` does not check either for redundant entries.

What do you think, @ShamrockLee ?

ShamrockLee replied:
> While I like the idea of going `pythonRelaxBuildDeps.foo = ">1, <2";` or such, especially if version constraints need to be (in part) programatically generated

Sounds great!

For such interface, are you planning to handle `pythonRelaxDeps.foo = null` as a way to "cancel" `foo`'s relaxation?

ShamrockLee replied:
@mweinelt What do you think about the attribute set interface?

mweinelt replied:
From experience, I mostly just remove the constraint instead of mutating it. We just need to deal with overzealous pinning here.

nbraud commented:
@ShamrockLee I requested a review, simply because you seem to be the most frequent commiters to `python/hooks`.

Feel free to drop the request or suggest someone more appropriate.

ShamrockLee commented:
IMHO, it seems like an enhancement that belongs to `pythonRelaxDepsHook`, since the `build-system` specification is essentially the same as the `dependency` specification in `pyproject.toml`.

How about naming it as, say, `pyprojectRelaxDepsHook`, so that we could gradually replace the current `pythonRelaxDepsHook`?

ShamrockLee commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -303,6 +305,9 @@ let
           pythonRelaxDepsHook
         ]
+        ++ optionals relaxBuildDeps [
+          pythonRelaxBuildDepsHook
```
Dynamically toggling `__structuredAttrs` based on the relaxing attributes would be very surprising. I prefer throwing an error if the hook is invoked an `__structuredAttrs` is false.

```suggestion
          (
            lib.throwIf
              (! finalAttrs.__structuredAttrs or false)
              "${finalAttrs.pname or finalAttrs.name}: pythonRelaxBuildDepsHook requires __structuredAttrs = true."
              pythonRelaxBuildDepsHook
          )
```

nbraud replied:
Fair.  :+1:

ShamrockLee commented on pkgs/development/interpreters/python/hooks/default.nix:
```diff
@@ -230,6 +230,29 @@ in
 
+  pythonRelaxBuildDepsHook = callPackage (
+    {
+      makePythonHook,
```
```suggestion
      makePythonHook,
      writeTextFile,
```

BTW, hope that the use of `packaging` and `tomlkit` won't cause circular dependency in the future.

nbraud commented:
> IMHO, it seems like an enhancement that belongs to `pythonRelaxDepsHook`, since the `build-system` specification is essentially the same as the `dependency` specification in `pyproject.toml`.
>
> How about naming it as, say, `pyprojectRelaxDepsHook`, so that we could gradually replace the current `pythonRelaxDepsHook`?

I didn't add it there, because `pythonRelaxDepsHook` functions in a completely-different (and far more generic) way, by editing the wheel.

I don't think it would even be possible to manipulate (runtime) dependencies reliably by modifying `pyproject.toml`, since build systems might get deps metadata from different sections of the file, or from somewhere else entirely.

Thanks for the review though, and sorry for the late reply: I've been unfortunately rather ill.

ShamrockLee commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -303,6 +305,14 @@ let
           pythonRelaxDepsHook
         ]
+        ++ optionals relaxBuildDeps [
+          (
+            lib.throwIf
+              (! finalAttrs.__structuredAttrs or false)
+              "${finalAttrs.pname or finalAttrs.name}: pythonRelaxBuildDepsHook requires __structuredAttrs = true."
+              pythonRelaxBuildDepsHook
+          )
```
This is an improved version of my previous suggestion.

The use of  `lib.extendDerivtaion` makes the assertion more gentle and allows filtering out `pythonRelaxBuildDepsHook` from `nativeBuildInputs` during overriding when `__structuredAttrs = false` (e.g., to replace it with another hook). In addition, this version of suggestion utilizes the fact that the attribute `__structuredAttrs` will always be set by `stdenv.mkDerivation`, and is thus always available in `finalAttrs`. It also uses `getName finalAttrs` to get the package name in a unified way.

```suggestion
          (extendDerivation (
            !finalAttrs.__structuredAttrs
            -> throw "${getName finalAttrs}: pythonRelaxBuildDepsHook requires __structuredAttrs = true."
          ) { } pythonRelaxBuildDepsHook)
```

ShamrockLee commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -221,6 +222,7 @@ let
         else
           "setuptools";
 
+      relaxBuildDeps = finalAttrs.pythonRelaxBuildDeps or [ ] != [ ] || finalAttrs.pythonRemoveBuildDeps or [ ] != [ ];
```
This fixes the formatting of my previous suggestion.

```suggestion
      relaxBuildDeps =
        finalAttrs.pythonRelaxBuildDeps or [ ] != [ ] || finalAttrs.pythonRemoveBuildDeps or [ ] != [ ];

```

ShamrockLee commented:
Thanks for the detailed explanation. I hope you're on the path to feeling better. Health comes first.

nbraud commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -363,6 +373,9 @@ let
         ++ nativeBuildInputs
         ++ build-system;
 
+      # pythonRelaxBuildDepsHook expects attrs in JSON
+      __structuredAttrs = relaxBuildDeps;
+
```
Oops, sorry I meant to remove it and forgot  :sweat:

nbraud commented:
Squashed fixup commit, removed leftover let-binding

nbraud commented:
Rebased on `master`, squashed-in 177851bca2d3e27ba1b603041512307b63662a49, fixed one last issue

nicoonoclaste committed:
pythonPackages.pythonRelaxBuildDepsHook: clarify and comment

Also fixed a minor bug: `pythonRelaxBuildDeps` was defaulting to `{}`, not `[]`

nicoonoclaste committed:
fixup! pythonPackages.pythonRelaxBuildDepsHook: init

nicoonoclaste committed:
fixup! pythonPackages.pythonRelaxBuildDepsHook: init

nicoonoclaste committed:
fixup! ansible: tidy-up using `pythonRelaxBuildDepsHook`

ShamrockLee commented on pkgs/development/interpreters/python/hooks/python-relax-build-deps-hook.py:
It would be great to format and lint this file with a formatter and a linter.

I don't know if there is a recommended Python formatter/linter for Nixpkgs development. I personally like [Ruff](https://github.com/astral-sh/ruff), which is both a Black-compatible formatter and a linter.

mweinelt replied:
Yes, please let's ruff all python things.
